adapters/beintoo/beintoo.go

```diff
@@ -0,0 +1,256 @@
+	}
+	
+	uriObj, _ := url.Parse(endpoint)
+	parameters := url.Values{}
+	
+	if testing {
+		
+		parameters.Add("t", "1000")
+		parameters.Add("ts", "2060541160")
+		uriObj.RawQuery = parameters.Encode()
+
+	return uriObj.String()
+	}
+	
+	parameters.Add("t", strconv.FormatInt(timeout, 10))
+	parameters.Add("ts", strconv.FormatInt(time.Now().Unix(), 10))
+	parameters.Add("src", "pbserver")
+
+	uriObj.RawQuery = parameters.Encode()
+
+	return uriObj.String()
+}
+
+func (a *BeintooAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("No Imps in Bid Request"),
+		}}
+	}
+
+	if errs := preprocess(request); errs != nil && len(errs) > 0 {
+		return nil, append(errs, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error in preprocess of Imp, err: %s", errs),
+		})
+	}
+
+	data, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Error in packaging request to JSON"),
+		}}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+	if request.Site != nil {
+		addHeaderIfNonEmpty(headers, "Referer", request.Site.Page)
+	}
+
+	url := buildEndpoint(a.endpoint, a.testing, request.TMax)
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     url,
+		Body:    data,
+		Headers: headers,
+	}}, errs
+}
+
+func unpackImpExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpBeintoo, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var BeintooExt openrtb_ext.ExtImpBeintoo
+	if err := json.Unmarshal(bidderExt.Bidder, &BeintooExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ignoring imp id=%s, invalid ImpExt", imp.ID),
+		}
+	}
+
+	tagIDValidation, err := strconv.ParseInt(BeintooExt.TagID, 10, 64)
+	if err != nil || tagIDValidation == 0 {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ignoring imp id=%s, invalid tagid must be a String of numbers", imp.ID),
+		}
+	}
+
+	
+
+	return &BeintooExt, nil
+}
+
+func buildImpBanner(imp *openrtb.Imp) error {
+	imp.Ext = nil
+
+	if imp.Banner == nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Request needs to include a Banner object"),
+		}
+	}
+
+	
+
+	if banner.W == nil && banner.H == nil {
+		if len(banner.Format) == 0 {
```

Comment:
Now that the `banner` copy was removed, the following errors appear because `banner` should be referenced as `imp.Banner` again.
```
# github.com/prebid/prebid-server/adapters/beintoo [github.com/prebid/prebid-server/adapters/beintoo.test]
./beintoo.go:135:5: undefined: banner
./beintoo.go:136:10: undefined: banner
./beintoo.go:141:13: undefined: banner
./beintoo.go:142:3: undefined: banner
./beintoo.go:143:3: undefined: banner
./beintoo.go:144:3: undefined: banner
./beintoo.go:145:16: undefined: banner
```

Reply:
Also not that the `Imp` object is local to your adapter, but `Imp.Banner` is a pointer to the shared `Banner` object. So you will need to make a copy of `Banner` and assign that to `Imp.Banner` if you want to make any changes to `Banner`.